Many small fixes to be able to compile the code with Visual Studio 2019

soeroesg opened the pull request:
Many small fixes to be able to compile the code with Visual Studio 2019

# CMake:
Compiler C++ standard setting in cross-platform way
Compiler-specific flags for GCC/Clang and MSVC
allow OpenCV 4 (but give preference for 3)
find_library(Boost) and link
find_library(OpenSSL) and link
add_subdirectory(DBoW2)
build STATIC libraries (.lib) on WIN32 because the code does not define any SHARED library (.dll) interface
we may need to use static CRT (/MT) too if Pangolin is compiled with its default /MT flag

# DBoW2:
link with Boost
allow OpenCV 4
changed hardcoded LIBRARY_OUTPUT_PATH to CMAKE_LIBRARY_OUTPUT_DIRECTORY

# g2o:
correctly add platform definitions like add_definitions(-DWINDOWS)
added missing g2o_RUNTIME_OUTPUT_DIRECTORY definition
fix matrix_operations.h to avoid MSVC crash https://github.com/RainerKuemmerle/g2o/issues/91
changed deprecated namespace std::tr1 to std
can only build static library on Windows because classes have no DLL API
--> consider upgrading to newer g2o release that has DLL API

# Header files:
changed stdint-gcc. to stdint.h, later removed as it is not needed apparently
removed unistd.h // TODO: is this needed for Linux at all? I believe it was for usleep() only
adde <algorithm> for std::max

# Source files:
changed usleep() to std::this_thread::sleep_for()
removed trailing white space at many places
fix allocation of dynamic two-dimensional array Distances[N][N] in MapPoint.cc, line line 372

# OpenCV
upgrade old OpenCV C constants to C++ constants
changed #include <opencv/cv.h> to #include <opencv2/features2d.hpp> in ORBExtractor.h
in PnPSolver.cpp added #include <opencv2/core/core_c.h> for the C API

soeroesg committed:
Many small fixes to be able to compile the code with Visual Studio 2019

# CMake:
Compiler C++ standard setting in cross-platform way
Compiler-specific flags for GCC/Clang and MSVC
allow OpenCV 4 (but give preference for 3)
find_library(Boost) and link
find_library(OpenSSL) and link
add_subdirectory(DBoW2)
build STATIC libraries (.lib) on WIN32 because the code does not define any SHARED library (.dll) interface
we may need to use static CRT (/MT) too if Pangolin is compiled with its default /MT flag

# DBoW2:
link with Boost
allow OpenCV 4
changed hardcoded LIBRARY_OUTPUT_PATH to CMAKE_LIBRARY_OUTPUT_DIRECTORY

# g2o:
correctly add platform definitions like add_definitions(-DWINDOWS)
added missing g2o_RUNTIME_OUTPUT_DIRECTORY definition
fix matrix_operations.h to avoid MSVC crash https://github.com/RainerKuemmerle/g2o/issues/91
changed deprecated namespace std::tr1 to std
can only build static library on Windows because classes have no DLL API
--> consider upgrading to newer g2o release that has DLL API

# Header files:
changed stdint-gcc. to stdint.h, later removed as it is not needed apparently
removed unistd.h // TODO: is this needed for Linux at all? I believe it was for usleep() only
adde <algorithm> for std::max

# Source files:
changed usleep() to std::this_thread::sleep_for()
removed trailing white space at many places
fix allocation of dynamic two-dimensional array Distances[N][N] in MapPoint.cc, line line 372

# OpenCV
upgrade old OpenCV C constants to C++ constants
changed #include <opencv/cv.h> to #include <opencv2/features2d.hpp> in ORBExtractor.h
in PnPSolver.cpp added #include <opencv2/core/core_c.h> for the C API

soeroesg committed:
MapPoint.cc Distances as vector of vectors instead of float**

soeroesg committed:
added option about CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS to be able to build SHARED libraries.

formatting in g2o CMakeLists.txt

soeroesg committed:
Merge branch 'master' of https://github.com/UZ-SLAMLab/ORB_SLAM3

runer112 commented on Examples/Monocular-Inertial/mono_inertial_tum_vi.cc:
```diff
@@ -196,9 +196,10 @@ int main(int argc, char **argv)
-                usleep((T-ttrack)*1e6); // 1e6
-
+            if (ttrack < T) {
+                long usec = static_cast<long>((T - ttrack) * 1e6);
+                std::this_thread::sleep_for(std::chrono::microseconds(usec));
```
Can these two lines be simplified as:
```
std::this_thread::sleep_for(std::chrono::seconds<double>(T - ttrack));
```

runer112 commented:
I performed my own mediocre Windows conversion before seeing this PR (whoops). May as well make use of the duplicated work to provide some review, comparing the changes I had to make to yours.

runer112 reviewed (commented):
Overall, I think this hits just about all of the necessary changes I found for building with MSVC. There were two notable ones you didn't seem to run into that I did, though:

1. [This one you commented on in the Mac support PR](https://github.com/UZ-SLAMLab/ORB_SLAM3/pull/131/files/c37b85ef2ea7bdb888ceba0f010c247f56b35439#diff-352386ec485eae566a99eeb18fbe8ea7R55). `const KeyFrame*` is definitely wrong, as that's really a pointer to a `const KeyFrame` (don't you love C/C++ declaration syntax?). I had to make the same change to `KeyFrame* const`, which is really a const key type (`KeyFrame*`) as necessary.
2. [This variable being declared as a bool](https://github.com/UZ-SLAMLab/ORB_SLAM3/blob/c0acf237bc0832f19a44676bdc5a2a074874dbe7/include/LoopClosing.h#L197). _LoopClosing.cc_ tries to increment this variable, and bool increment is definitely not a standard-supported operation. Judging by how it was used and the fact that the name includes "Idx", my fix was to change the type to `unsigned int`.

soeroesg commented on CMakeLists.txt:
```diff
@@ -5,125 +5,155 @@ IF(NOT CMAKE_BUILD_TYPE)
+  set(ORB_SLAM3_LIB_TYPE STATIC)
+  # or we have to export all symbols from all SHARED libraries
+  #set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
+  # Note that the ideal solution would be to add an export API to the relevant classes.
```
I would do the DLL interface in a separate merge request later because I also expect a lot of problems with the outdated g2o library. We should upgrade to a new version of g2o first, which already comes with a DLL interface

soeroesg replied:
About the first point, we should ask the original authors (@richard-elvira, @ccamposm) what is the intention here. It seems the key must be constant in the map, but I also think the Keyframe should not change, so maybe `const KeyFrame *const` is the correct entry here.

I correct the second point, thanks for pointing it out! I change it to std::size_t

soeroesg replied:
btw, I have Eigen 3.3.7, MSVC 16.7, and it has no problem with the original `const KeyFrame*`

another question to this: should be the key type in the first line (50) be the same as the key type in the second line (51)?

soeroesg commented on Examples/Monocular-Inertial/mono_inertial_euroc.cc:
```diff
@@ -193,8 +193,10 @@ int main(int argc, char *argv[])
-            if(ttrack<T)
-                usleep((T-ttrack)*1e6); // 1e6
+            if (ttrack < T) {
+                long usec = static_cast<long>((T - ttrack) * 1e6);
+                std::this_thread::sleep_for(std::chrono::microseconds(usec));
```
yes, we could shorten it, but I prefer to explicitly see the casting from double to long, this is why I kept it in separate lines

soeroesg commented on include/System.h:
```diff
@@ -22,12 +22,15 @@
+#include <stdlib.h>
+#include <string>
+#include <thread>
+#ifndef WIN32
+  #include <unistd.h>
+#endif
```
will do in my next commit

soeroesg commented on Thirdparty/DBoW2/CMakeLists.txt:
```diff
@@ -1,19 +1,27 @@
 cmake_minimum_required(VERSION 2.8)
```
the root cmakelist sets the c++ standard globally, so i think it is not necessary here

soeroesg replied:
the COMPILEDWITHC11 flag is not used in DBoW2 anyway

soeroesg committed:
removed unistd.h

soeroesg commented on Thirdparty/DBoW2/CMakeLists.txt:
```diff
@@ -24,17 +32,33 @@ set(SRCS_DUTILS
 
+# add Boost
+find_package(Boost REQUIRED
+  COMPONENTS filesystem system serialization)
```
you're right, I remove filesystem and system

soeroesg committed:
changed type of mnFullBAIdx from bool to std::size_t

soeroesg committed:
CMakeLists.txt: removed unnecessary boost::filesystem and boost::system

soeroesg committed:
CMakeLists.txt of g2o:
set the G2O_SOURCE_DIR variable which was unset before and therefore the configuration of config.h.in may have failed.

trzy commented:
Hey @soeroesg , I'm still seeing a crash caused by g2o's sparse_block_matrix.hpp (which in turn calls internal::axp() from matrix_operations.h) with Visual Studio 2019. Your fix modifies the syntax of the specialized templates but I believe line 277 of sparse_block_matrix.hpp is calling the generic template on line 35 of matrix_operations.h. 

Using Visual Studio 2019 Community v16.8.2 here. 

Any suggestions would be much appreciated :)

antithing commented:
@trzy try making these changes in matrix_operations.h

```
 template<typename MatrixType>
    inline void axpy(const MatrixType& A, const Eigen::Map<const Eigen::VectorXd>& x, int xoff, Eigen::Map<Eigen::VectorXd>& y, int yoff)
    {
      y.segment<MatrixType::RowsAtCompileTime>(yoff) += A * x.segment<MatrixType::ColsAtCompileTime>(xoff);
    }

    template<int t>
    inline void axpy(const Eigen::Matrix<double, Eigen::Dynamic, t>& A, const Eigen::Map<const Eigen::VectorXd>& x, int xoff, Eigen::Map<Eigen::VectorXd>& y, int yoff)
    {
      y.segment(yoff, A.rows()) += A * x.segment<Eigen::Matrix<double, Eigen::Dynamic, t>::ColsAtCompileTime>(xoff);
    }

    template<>
	//inline void axpy(const Eigen::MatrixXd& A, const Eigen::Map<const Eigen::VectorXd>& x, int xoff, Eigen::Map<Eigen::VectorXd>& y, int yoff)
	inline void axpy<Eigen::MatrixXd>(const Eigen::MatrixXd& A, const Eigen::Map<const Eigen::VectorXd>& x, int xoff, Eigen::Map<Eigen::VectorXd>& y, int yoff)
    {
      y.segment(yoff, A.rows()) += A * x.segment(xoff, A.cols());
    }

    template<typename MatrixType>
    inline void atxpy(const MatrixType& A, const Eigen::Map<const Eigen::VectorXd>& x, int xoff, Eigen::Map<Eigen::VectorXd>& y, int yoff)
    {
      y.segment<MatrixType::ColsAtCompileTime>(yoff) += A.transpose() * x.segment<MatrixType::RowsAtCompileTime>(xoff);
    }

    template<int t>
    inline void atxpy(const Eigen::Matrix<double, Eigen::Dynamic, t>& A, const Eigen::Map<const Eigen::VectorXd>& x, int xoff, Eigen::Map<Eigen::VectorXd>& y, int yoff)
    {
      y.segment<Eigen::Matrix<double, Eigen::Dynamic, t>::ColsAtCompileTime>(yoff) += A.transpose() * x.segment(xoff, A.rows());
    }

    template<>
	//inline void atxpy(const Eigen::MatrixXd& A, const Eigen::Map<const Eigen::VectorXd>& x, int xoff, Eigen::Map<Eigen::VectorXd>& y, int yoff)
	inline void atxpy<Eigen::MatrixXd>(const Eigen::MatrixXd& A, const Eigen::Map<const Eigen::VectorXd>& x, int xoff, Eigen::Map<Eigen::VectorXd>& y, int yoff)
    {
      y.segment(yoff, A.cols()) += A.transpose() * x.segment(xoff, A.rows());
    }
```

trzy commented:
@antithing Unfortunately that doesn't work for me. I get the same error.

My nasty workaround was to define two macros:

```
#define _AXPY(MatrixType,A,x,xoff,y,yoff)y.segment<MatrixType::RowsAtCompileTime>(yoff) += (A) * (x).segment<MatrixType::ColsAtCompileTime>(xoff)
#define _ATXPY(MatrixType,A,x,xoff,y,yoff)y.segment<MatrixType::ColsAtCompileTime>(yoff) += (A).transpose() * (x).segment<MatrixType::RowsAtCompileTime>(xoff)
```

And then I simply have to replace every invocation of g2o::internal::atxpy(). For example, in sparse_block_matrix.hpp:

```
        // destVec += *a * srcVec (according to the sub-vector parts)
        //internal::axpy(*a, srcVec, srcOffset, destVec, destOffset);
        _AXPY(MatrixType, *a, srcVec, srcOffset, destVec, destOffset);
        if (destOffset < srcOffset)
          //internal::atxpy(*a, srcVec, destOffset, destVec, srcOffset);
          _ATXPY(MatrixType, *a, srcVec, destOffset, destVec, srcOffset);
      }
```

I am still very interested in understanding why my very recent version of VS2019 chokes on the code you posted.

soeroesg committed:
Merge remote-tracking branch 'upstream/master'

soeroesg committed:
fix compilation error with OpenCV 3.2

soeroesg committed:
fix of the previous fix

soeroesg committed:
fix of fix of fix...

soeroesg committed:
fixed problem with library output paths on Windows

soeroesg committed:
added build_win.cmd

soeroesg committed:
save trajectory at the end of all ROS examples

soeroesg committed:
CMake changes:
removed global includes
library linking PRIVATE
added CMAKE_DEBUG_POSTFIX
added /bigobj MSVC flag
removed (the anyway unused) OpenSSL dependency
formatting, reordering

ignore bin folder

soeroesg committed:
reverted changes in matrix_operations.h, not needed anymore

soeroesg commented:
Interestingly, the problem disappeared for me after upgrading to Visual Studio 2019 16.10.
What's more, the changes to the template are not even needed anymore, the old code seems to work fine. So I reverted those lines in matrix_operations.h

soeroesg committed:
made OpenCV, Eigen, Pangolin dependencies PUBLIC in cmake as they are indirectly used by the Examples

soeroesg committed:
added win script to run example

soeroesg committed:
public link DBoW2 with OpenCV and Boost

soeroesg committed:
public link g2o with Eigen; configure G2O_SHARED_LIBS; do not alter library output dir

soeroesg committed:
cmake print dependency library versions; link with Backward in Debug mode

soeroesg committed:
build scripts can specify path to installed dependencies; added example scripts to run euroc MH01

soeroesg committed:
Merge branch 'UZ-SLAMLab:master' into master

jpolin commented on CMakeLists.txt:
```diff
@@ -1,131 +1,192 @@
--lboost_serialization
--lcrypto
-)
+# TODO: OpenSSL is used only in System.cc #include <openssl/md5.h>, but then MD5 is never used.
```
FYI this is what I was worried about. Still playing around with it, but running out of ideas. Open to ideas...

jpolin replied:
sorry just needed to screenshot something, ignore me :)
